### ISSUE TITLE
Logstash output?

### ISSUE THREAD

**Issue by barryw** [elastic/apm-server#664]
Does the roadmap include using logstash as an output target?

**Comment by graphaelli** [comment]
It is on the roadmap but without a solid timeline at the moment.  We'd be curious to hear your use case, in particular any reasons you might use logstash instead of feeding elasticsearch directly.  Also, do you plan to use sourcemapping capabilities?

**Comment by barryw** [comment]
What I'm trying to do is feed all of my service APM data into logstash so that I can do geoip and user_agent lookups before forwarding to elasticsearch. I'm gathering data from several services (node & python), and not all have IP/useragent data. I thought I could get away with using an elasticsearch ingest pipeline, but it's not flexible enough to deal with missing IP/useragent data.

Ideally, I'd love to funnel all of my data (beats, logspout & apm) through logstash since it's just incredibly flexible.

**Comment by vbohata** [comment]
+1 for logstash output. Use case: large log management solution built on logstash, kafka, elasticsearch, kibana. It is not wanted to write to elasticsearch directly. Logstash + Kafka brings possibilities to do CPU intensive data parsing, works as a protection/throttling for ES in case some client/agent start to send too many data to elasticsearch, extends high availability possibilities (no problem to have ES completely down for some time without loosing data)...

**Comment by ball-hayden** [comment]
I posted another use case to the ES forum today (before I came across this issue):
https://discuss.elastic.co/t/apm-server-offline-retention-synchronization/137155

To summarise: the Elastic ecosystem has a nice way of supporting asynchronous transfer of data via Logstash. It would be nice to see that option supported for APM 😄 

(and yes, I would be wanting to use source mapping features)

**Comment by graphaelli** [comment]
While this is currently in development, some bits are already in place, namely apm-server speaking the beats protocol to logstash.  If you're feeling very brave and not looking for official guarantees or support, apm-server can be configured to output to logstash with:

```
# disable elasticsearch output
output.elasticsearch.enabled=false
# configure logstash output
output.logstash.enabled=true
output.logstash.hosts=["logstash:5044"]
```

On the logstash side:
```
input {
  beats {
    id => "apm-server"
    port => 5044
  }
}

output {
  elasticsearch {
    hosts => ["elasticsearch:9200"]
    index => "apm-%{[beat][version]}-%{+YYYY.MM.dd}"
  }
}
```

Note that sourcemaps will not be applied as the elasticsearch output is currently used as the sourcemap storage location - decoupling these is part of logstash work that's planned.

**Comment by graphaelli** [comment]
Closed by #1103 -Documentation for this is now available: https://www.elastic.co/guide/en/apm/server/master/logstash-output.html

Watch #1114 for details on sourcemap-specific details.
